### DISCUSSION TITLE
Powerwall data on Tidbyt

### DISCUSSION THREAD

**Discussion by dsf01** [jasonacox/Powerwall-Dashboard#189] [Ideas]
What would be the effort to display powerwall data on a Tidbyt device?  It’s a small standalone display that allows custom app deployment.

https://tidbyt.com/

https://tidbyt.dev/

**Comment by BJReplay** [comment]
The data made available from pypowerwall would be able to be displayed on this device, but everything would need to be written, more or less from scratch, for the display.  You wouldn't be able to borrow from the grafana work, because this device won't be able to render anything from grafana (grafana needs a modern browser).

I've got this little graph displaying in my grafana dashboard, and it would be relatively simple for a developer to replicate something like this (from scratch) in starlark to display, for example.

![image](https://user-images.githubusercontent.com/37993507/220767238-4f80c5f2-42e1-4327-b9cb-c3aac7785968.png).

I'm semi-retired; if someone wants to buy one and send it to me in Australia, I'll develop the basics and a display and happily add it to this project (but I'm not that motivated to buy one myself 🤣).

**Reply by jasonacox** [reply]
Wow, that's cool @dsf01 !  

What @BJReplay says is correct.  Grafana won't work on this, but pypowerwall could provide the data you need.  Keep in mind pypowerwall only provides "current state," not history.  To get history you either need to store that data in the device itself or you would pull it from the InfluxDB that the dashboard project uses.

I noticed this: https://github.com/tidbyt/community/tree/main/apps/solaredgemonitor

![image](https://user-images.githubusercontent.com/836718/220820797-59d9e676-20ff-4bac-aab6-31f95ff139a0.png)

If you wanted something like that, it would be a simple code change to point to pypowerwall. You could grab all of that data with one call to http://pypowerwall:8675/csv which will produce key power data in CSV format: (grid, home, solar, battery, batterylevel):

```csv
# grid,home,solar,battery,batterylevel
17.00,1701.50,10.00,1670.00,75.48
```

Or you can pull more details from other pypowerwall APIs (see some options here:  https://github.com/jasonacox/pypowerwall/blob/main/proxy/HELP.md).

It looks like they have a software emulator for developers (or someone can send a unit to @BJReplay 😂 ) and we could get it to work with pypowerwall, and submit a PR to the Tidbyt github: https://github.com/tidbyt/community to publish the app (see [here](https://tidbyt.dev/docs/publish/publishing-apps)).

**Reply by BJReplay** [reply]
> or someone can send a unit to @BJReplay 😂

I thought it was worth putting it out there 😜

**Comment by dsf01** [comment]
Would someone(s) like to split the cost to fund a dev unit as a thank you for someone willing to take on the challenge?

**Reply by BJReplay** [reply]
@dsf01 I'm not sure how well this will work - I'm not saying don't work out if you want to try, but I'm saying that it might take a fair bit of set up.  From what I read [here](https://github.com/tidbyt/pixlet/blob/dc0bd878183177fd1315045c727d7c5ff4207d1b/docs/authoring_apps.md)

> The Tidbyt displays tiny 64x32 animations and images. It keeps a local cache of the apps that are installed on it.
> 
> Each Tidbyt regularly sends heartbeats to the Tidbyt cloud, announcing what it has cached. The Tidbyt cloud decides which app needs to be rendered next, and executes the appropriate Starlark script. It encodes the result as a [WebP](https://developers.google.com/speed/webp) image, and sends it back to the device.
> 
> To mimic how we host apps internally, pixlet render executes the Starlark script and pixlet push pushes the resulting WebP to your Tidbyt.
> 

What this means is that for a published app, it wouldn't be possible just to configure the IP address or hostname of the server running pypowerwall, and code up an app and publish it, you'd actually have to set up pixlet, and have that running locally, serving up the information to the Tidbyt display locally (because there is no way - unless you open your pypowerwall docker service to the internet via a reverse proxy - that would be a bad idea - for the Tidbyt cloud to access the data to push it to the device.

Now, because I'm a curious person, I've got the Hello World app up and running, but I'm not certain that this is doable, yet, without getting pixlet up and running on your Powerwall-Gateway server as well.  But I'm willing to give it a crack 😉

**Comment by BJReplay** [comment]
Further to this: See this discussion: https://discuss.tidbyt.com/t/self-hosted-tidbyt-server/330/4

Unless you can open up your data to the Tidbyt cloud, I suspect this won't fly.

That said, because it is Friday and beer o'clock...

![image](https://user-images.githubusercontent.com/37993507/221120134-a5811a68-835f-4eb0-b889-33c63dbc2d15.png)

It's effectively static.  No way to push updates as far as I can tell.

My data is out there on the web via a reverse proxy, so I could do this (but I'm not going to).

If people are prepared to reverse proxy their pypowerwall data (specifically /soe and /aggregates I suspect I could build an app that could be published to the Tidbyt cloud, and thus enable a live display, but I suspect that a local server won't allow for live updates to a local device, unfortunately.

**Reply by dsf01** [reply]
Thank you for the exhaustive research.  I appreciate the effort @BJReplay

**Reply by BJReplay** [reply]
No worries.  You've got to something while drinking beer and supposedly doing some work on a Friday afternoon.

The other thought I had overnight was anyone could probably poll the Tesla servers in the cloud for this data.  The config that would be required would be a Tesla API refresh token (which you could obtain from one of the various token apps either on the web or on phone apps either Android or iPhone).  That would be available to anyone, and provide a live feed - just like the official Tesla app on your phone uses.

So, it's definitely doable.

**Comment by dsf01** [comment]
There was an app published today to archive this -

https://github.com/tidbyt/community/pull/1226

Unfortunately I don’t know my RefreshToken and SiteID.
